### PR TITLE
Give user a better feedback when commands are executed

### PR DESCRIPTION
**Pull request by jakolehm**
- use spinner on all commands that modify system (spinner waits state change when possible)
- improve deploy tracking

### CODE REVIEW

**Comment by matti** [comment]
I'll give this a spin tomorrow morning

**Comment by matti** [comment]
@jakolehm would it be possible to have automated docker builds of feature branches like these for easier testing (with contena for example)

**Comment by matti** [comment]
We install kontena cli with Gemfile for the project so the old gemset will conflict with my symlinked "devtena"...

File: cli/lib/kontena/cli/apps/restart_command.rb [hunk 1]
```diff
@@ -27,10 +27,11 @@ def execute
...
       services.each do |service_name, opts|
         if service_exists?(service_name)
-          puts "restarting #{prefixed_name(service_name)}"
-          restart_service(token, prefixed_name(service_name))
+          ShellSpinner "sending restart signal to #{service_name.colorize(:cyan)} " do
+            restart_service(token, prefixed_name(service_name))
+          end
         else
-          puts "No such service: #{service_name}".colorize(:red)
+          STDERR.puts "WARNING: no such service: #{service_name}".colorize(:yellow)
```
Comment (matti) [inline cli/lib/kontena/cli/apps/restart_command.rb]: I don't think this is a "warning", isn't this "I have not done anything, aboring the whole thing", I think it should be in red and a "ERROR" + give exit code other than 0 (and use the abort() method?)

`kontena app restart && echo "successfully restarted!"` should not happen if the restart fails

File: cli/lib/kontena/cli/apps/remove_command.rb [hunk 2]
```diff
@@ -41,17 +42,14 @@ def remove_services(services)
...
-        ShellSpinner " " do
+        ShellSpinner "removing #{name.colorize(:cyan)}" do
           service = get_service(token, prefixed_name(name)) rescue nil
           if(service)
             delete_service(token, prefixed_name(name))
             unless async
               wait_for_delete_to_finish(service)
             end
             deleted_services << name
-          else
```
Comment (matti) [inline cli/lib/kontena/cli/apps/remove_command.rb]: Why is else removed? Shouldn't this be an error?

File: cli/lib/kontena/cli/apps/start_command.rb [hunk 3]
```diff
@@ -28,10 +28,11 @@ def execute
...
       services.each do |service_name, opts|
         if service_exists?(service_name)
-          puts "starting #{prefixed_name(service_name)}"
-          start_service(token, prefixed_name(service_name))
+          ShellSpinner "starting #{service_name.colorize(:cyan)} " do
+            start_service(token, prefixed_name(service_name))
+          end
         else
-          puts "No such service: #{service_name}".colorize(:red)
+          STDERR.puts "WARNING: no such service: #{service_name}".colorize(:yellow)
```
Comment (matti) [inline cli/lib/kontena/cli/apps/start_command.rb]: See the previous comment for all of these

File: cli/lib/kontena/cli/grids/create_command.rb [hunk 4]
```diff
@@ -16,10 +16,17 @@ def execute
...
+      if initial_size == 1
+        STDERR.puts "WARNING: --initial-size=1 is only recommended for test/dev usage".colorize(:yellow)
+      end
+      ShellSpinner "creating #{name.colorize(:cyan)} grid " do
+        grid = client(token).post('grids', payload)
+      end
       if grid
-        self.current_grid = grid
-        puts "Using grid: #{grid['name'].cyan}"
+        ShellSpinner "switching scope to #{name.colorize(:cyan)} grid " do
```
Comment (matti) [inline cli/lib/kontena/cli/grids/create_command.rb]: NICE!

File: cli/lib/kontena/cli/registry/remove_command.rb [hunk 5]
```diff
@@ -8,11 +8,14 @@ def execute
       require_api_url
       token = require_token
       confirm unless forced?
+      name = 'registry'
 
-      registry = client(token).get("services/#{current_grid}/registry") rescue nil
-      abort("Docker Registry service does not exist") if registry.nil?
+      registry = client(token).get("services/#{current_grid}/#{name}") rescue nil
+      abort("#{name.colorize(:cyan)} service does not exist") if registry.nil?
```
Comment (matti) [inline cli/lib/kontena/cli/registry/remove_command.rb]: (see previous comments) To unify things CLI should use the abort if something does not exist, e.g. the scope of the command can not be performed. abort() should output red "ERROR:" prefix

File: cli/lib/kontena/cli/services/remove_command.rb [hunk 6]
```diff
@@ -13,7 +13,17 @@ def execute
...
 
-      result = client(token).delete("services/#{parse_service_id(name)}")
+      ShellSpinner "removing service #{name.colorize(:cyan)} " do
+        client(token).delete("services/#{parse_service_id(name)}")
+        removed = false
+        until removed == true
+          begin
+            client(token).get("services/#{parse_service_id(name)}")
+          rescue Kontena::Errors::StandardError => exc
+            removed = true if exc.status == 404
```
Comment (matti) [inline cli/lib/kontena/cli/services/remove_command.rb]: else missing, should use abort()

File: cli/lib/kontena/cli/services/secrets/link_command.rb [hunk 7]
```diff
@@ -12,13 +12,15 @@ class LinkCommand < Clamp::Command
...
       token = require_token
-      result = client(token).get("services/#{parse_service_id(name)}")
-      secrets = result['secrets']
-      secrets << parse_secrets([secret])[0]
-      data = {
-        secrets: secrets
-      }
-      client(token).put("services/#{parse_service_id(name)}", data)
+      ShellSpinner "linking #{secret.colorize(:cyan)} from Vault to #{name.colorize(:cyan)} " do
+        result = client(token).get("services/#{parse_service_id(name)}")
```
Comment (matti) [inline cli/lib/kontena/cli/services/secrets/link_command.rb]: Fails ugly if vault the service does not exists:
```
linking ff from Vault to asdf ... fail
/Users/mpa/dev/kontena/cli/lib/kontena/client.rb:209:in `handle_error_response': {"error":"Not found"} (Exception)
	from /Users/mpa/dev/kontena/cli/lib/kontena/client.rb:43:in `get'
``

File: cli/lib/kontena/cli/services/secrets/link_command.rb [hunk 8]
```diff
@@ -12,13 +12,15 @@ class LinkCommand < Clamp::Command
...
+        secrets = result['secrets']
+        secrets << parse_secrets([secret])[0]
+        data = {
+          secrets: secrets
+        }
+        client(token).put("services/#{parse_service_id(name)}", data)
```
Comment (matti) [inline cli/lib/kontena/cli/services/secrets/link_command.rb]: Not maybe part of this PR's scope, but I find it odd that you can link non-existing secrets..?

File: cli/lib/kontena/cli/services/services_helper.rb [hunk 9]
```diff
@@ -214,13 +214,25 @@ def deploy_service(token, service_id, data)
...
+              deployed = containers.size == desired_count && containers.all?{ |c|
+                DateTime.parse(c['created_at']) >= updated_at
+              }
               sleep 1
             end
           end
+
+          deployed
+        rescue Timeout::Error
+          raise Kontena::Errors::StandardError.new(500, 'deploy timed out')
```
Comment (matti) [inline cli/lib/kontena/cli/services/services_helper.rb]: Use abort() instead?

File: cli/lib/kontena/cli/vpn/create_command.rb [hunk 10]
```diff
@@ -36,13 +37,28 @@ def execute
...
-      puts "Use 'kontena vpn config' to fetch OpenVPN client config to your machine (it takes a while until config is ready)."
+      puts "#{name.colorize(:cyan)} service is now started (udp://#{vpn_ip}:1194)."
+      puts "use 'kontena vpn config' to fetch OpenVPN client config to your machine."
     end
 
+    def wait_for_configuration_to_finish(token)
+      finished = false
+      payload = {cmd: ['/usr/local/bin/ovpn_getclient', 'KONTENA_VPN_CLIENT']}
+      until finished
+        sleep 30
```
Comment (matti) [inline cli/lib/kontena/cli/vpn/create_command.rb]: Never times out? Isn't sleep 30 a bit long?

**Comment by jakolehm** [comment]
Rebased against master and refactored/unified spinner & error messages

File: cli/lib/kontena/cli/apps/remove_command.rb (hunk 2 above)
Comment (jakolehm) [inline cli/lib/kontena/cli/apps/remove_command.rb]: Refactored it so that warning is displayed to user

File: cli/lib/kontena/cli/apps/restart_command.rb (hunk 1 above)
Comment (jakolehm) [inline cli/lib/kontena/cli/apps/restart_command.rb]: I'm not completely sold on the idea that "app" commands should exit with error if some service does not exist. At least these kind of changes should be done in a separate pr. @nevalla what do you think?

**Comment by jakolehm** [comment]
@nevalla ping

**Comment by nevalla** [comment]
Looks good!
